ctru-rs/examples/camera-image.rs

```diff
@@ -0,0 +1,202 @@
+use ctru::console::Console;
+use ctru::gfx::{Screen, Side};
+use ctru::services::cam::{
+    Cam, CamContext, CamOutputFormat, CamPort, CamSelect, CamShutterSoundType, CamSize,
+};
+use ctru::services::hid::KeyPad;
+use ctru::services::{Apt, Hid};
+use ctru::Gfx;
+
+const WIDTH: i16 = 400;
+const HEIGHT: i16 = 240;
+const SCREEN_SIZE: u32 = 192000; // WIDTH * HEIGHT * 2;
+const BUF_SIZE: usize = 384000; // SCREEN_SIZE * 2;
+
+const WAIT_TIMEOUT: i64 = 300000000;
+
+fn main() {
+    ctru::init();
+
+    let apt = Apt::init().expect("Failed to initialize Apt service.");
+    let hid = Hid::init().expect("Failed to initialize Hid service.");
+    let gfx = Gfx::init().expect("Failed to initialize GFX service.");
+
+    gfx.top_screen.borrow_mut().set_double_buffering(true);
+    gfx.bottom_screen.borrow_mut().set_double_buffering(false);
+
+    let _console = Console::init(gfx.bottom_screen.borrow_mut());
+
+    let mut keys_down;
+
+    println!("Initializing camera");
+
+    let mut cam = Cam::init().expect("Failed to initialize CAM service.");
+
+    cam.set_size(
+        CamSelect::OUT1_OUT2,
+        CamSize::CTR_TOP_LCD,
+        CamContext::A,
+    )
+    .expect("Failed to set camera size");
+    cam.set_output_format(
+        CamSelect::OUT1_OUT2,
+        CamOutputFormat::RGB_565,
+        CamContext::A,
+    )
+    .expect("Failed to set camera output format");
+
+    cam.set_noise_filter(CamSelect::OUT1_OUT2, true)
+        .expect("Failed to enable noise filter");
+    cam.set_auto_exposure(CamSelect::OUT1_OUT2, true)
+        .expect("Failed to enable auto exposure");
+    cam.set_auto_white_balance(CamSelect::OUT1_OUT2, true)
+        .expect("Failed to enable auto white balance");
+
+    cam.set_trimming(CamPort::CAM1, false)
+        .expect("Failed to disable trimming for Cam Port 1");
+    cam.set_trimming(CamPort::CAM2, false)
+        .expect("Failed to disable trimming for Cam Port 2");
+
+    let mut buf = [0u8; BUF_SIZE];
+
+    gfx.flush_buffers();
+    gfx.swap_buffers();
+    gfx.wait_for_vblank();
+
+    println!("\nPress R to take a new picture");
+    println!("Press Start to exit to Homebrew Launcher");
+
+    while apt.main_loop() {
+        hid.scan_input();
+        keys_down = hid.keys_down();
+
+        if keys_down.contains(KeyPad::KEY_START) {
+            break;
+        }
+
+        if keys_down.contains(KeyPad::KEY_R) {
+            println!("Capturing new image");
+            take_picture(&mut cam, &mut buf);
+        }
+
+        let img = convert_image_to_rgb8(buf, 0, 0, WIDTH as usize, HEIGHT as usize);
+
+        unsafe {
+            gfx.top_screen
+                .borrow_mut()
+                .get_raw_framebuffer(Side::Left)
+                .ptr
+                .copy_from(img.as_ptr(), img.len());
```

Comment:
Isn't `img` the size of two screens? Therefore writing all of its contents into a framebuffer the size of one screen would mean you're writing past the buffer's limits?

Reply:
I'm going to change `BUF_SIZE` to represent a single image. I mentioned this before in the comment thread about the conversion function. The reason BUF_SIZE was the size of multiple was for 3D as the cameras would take two images. In that case, writing to the frame buffer would be a little different as well I believe

Reply:
After playing around with this, it seems like the rgb8 format for the frame buffer may have something to do with how it's read as two images. I tried halving `BUF_SIZE` but it causes an out of bounds exception when converting. I'd have to look into 3D more probably to understand how that's actually being sent to the buffer cause it may not be as straight forward as sending the two halves.